tokio-util/tests/spawn_pinned.rs

```diff
@@ -0,0 +1,186 @@
+#[tokio::test]
+async fn can_spawn_multiple_futures() {
+    let pool = task::LocalPoolHandle::new(2);
+
+    let join_handle1 = pool.spawn_pinned(|| {
+        let local_data = Rc::new("test1");
+        async move { local_data.to_string() }
+    });
+    let join_handle2 = pool.spawn_pinned(|| {
+        let local_data = Rc::new("test2");
+        async move { local_data.to_string() }
+    });
+
+    assert_eq!(join_handle1.await.unwrap(), "test1");
+    assert_eq!(join_handle2.await.unwrap(), "test2");
+}
+
+/// A panic in the spawned task causes the join handle to return an error.
+/// But, you can continue to spawn tasks.
+#[tokio::test]
+async fn task_panic_propagates() {
+    let pool = task::LocalPoolHandle::new(1);
+
+    let join_handle = pool.spawn_pinned(|| async {
+        panic!("Test panic");
+    });
+
+    let result = join_handle.await;
+    assert!(result.is_err());
+    let error = result.unwrap_err();
+    assert!(error.is_panic());
+
+    // Trying again with a "safe" task still works
+    let join_handle = pool.spawn_pinned(|| async { "test" });
+    let result = join_handle.await;
+    assert!(result.is_ok());
+    assert_eq!(result.unwrap(), "test");
+}
+
+/// A panic during task creation causes the join handle to return an error.
+/// But, you can continue to spawn tasks.
+#[tokio::test]
+async fn callback_panic_does_not_kill_worker() {
+    let pool = task::LocalPoolHandle::new(1);
+
+    let join_handle = pool.spawn_pinned(|| {
+        panic!("Test panic");
+        #[allow(unreachable_code)]
+        async {}
+    });
+
+    let result = join_handle.await;
+    assert!(result.is_err());
+    let error = result.unwrap_err();
+    assert!(error.is_panic());
+
+    // Trying again with a "safe" callback works
+    let join_handle = pool.spawn_pinned(|| async { "test" });
+    let result = join_handle.await;
+    assert!(result.is_ok());
+    assert_eq!(result.unwrap(), "test");
+}
+
+/// Canceling the task via the returned join handle cancels the spawned task
+/// (which has a different, internal join handle).
+/// This test is loosely based off of `test_abort_wakes_task_3964` in
+/// `tokio/tests/test_abort.rs`
+#[tokio::test]
+async fn task_cancellation_propagates() {
+    let pool = task::LocalPoolHandle::new(1);
+    let notify_dropped = Arc::new(());
+    let weak_notify_dropped = Arc::downgrade(&notify_dropped);
+
+    let join_handle = pool.spawn_pinned(|| async move {
+        // Move the Arc into the task
+        let _notify_dropped = notify_dropped;
+        println!("task started");
+
+        // Sleep for a while (should be canceled before it elapses)
+        tokio::time::sleep(Duration::new(100, 0)).await
+    });
+
+    // Wait for inner task to sleep
+    tokio::time::sleep(Duration::from_millis(10)).await;
+
+    join_handle.abort();
+
+    // Wait for inner task to abort
+    tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
Sleep based tests are pretty bad for the overall performance of our test suite. This test alone will take hundreds of times longer than any other test in this file, and 10 ms sleeps really do add up. We have _a lot_ of tests in Tokio.

You can have the task send a message when it starts running and then `.await` the `JoinHandle` afterwards to wait for its destructor to run.